File: app/lib/chat/thread_key.dart
// Canonical thread key for message timelines (must match web + backend).
//
// Format:
// - Account: u:{userId} (logged in) or o:{offlineUserId} (offline).
// - 1:1 device: {account}|d1:{min}|d2:{max} (lexicographic min/max of the two device ids).
// - S3 cloud entry: {account}|kind:s3_cloud.
// - Legacy broadcast: {account}|kind:legacy_broadcast.

// Same value as s3VirtualDeviceId in device_provider.dart and S3_VIRTUAL_DEVICE_ID on web.
const kS3VirtualDeviceId = '__s3_cloud__';

const threadKindS3Cloud = 's3_cloud';
const threadKindLegacyBroadcast = 'legacy_broadcast';

String accountPartLoggedIn(String userId) => 'u:$userId';

String accountPartOffline(String offlineUserId) => 'o:$offlineUserId';

String threadKeyOneToOne(
  String accountPart,
  String deviceIdA,
  String deviceIdB,
) {
  final a = deviceIdA.compareTo(deviceIdB) <= 0 ? deviceIdA : deviceIdB;
  final b = deviceIdA.compareTo(deviceIdB) <= 0 ? deviceIdB : deviceIdA;
  return '$accountPart|d1:$a|d2:$b';
}

String threadKeyS3Cloud(String accountPart) =>
    '$accountPart|kind:$threadKindS3Cloud';

String threadKeyLegacyBroadcast(String accountPart) =>
    '$accountPart|kind:$threadKindLegacyBroadcast';

/// Thread for the current sidebar selection (peer or S3 virtual).
String threadKeyForPeerSelection({
  required String accountPart,
  required String myDeviceId,
  required String selectedPeerId,
}) {
  if (selectedPeerId == kS3VirtualDeviceId) {
    return threadKeyS3Cloud(accountPart);
  }
  return threadKeyOneToOne(accountPart, myDeviceId, selectedPeerId);
}

/// Derives [threadKey] for persistence when optional [explicitThreadKey] is absent.
///
/// - If [toDeviceId] is non-null and not S3 virtual: canonical pair (from, to).
/// - Else if [toDeviceId] is null and [fromDeviceId] != [myDeviceId]: treat as incoming
///   without explicit recipient → 1:1 with sender and this device.
/// - Else: legacy broadcast bucket (old outbound without [toDeviceId]).
/// Keep an explicit key only when it belongs to [accountPart]; otherwise
/// derive locally so guest IM envelopes from another device still land in
/// this device's 1:1 thread.
String? localExplicitThreadKey(String accountPart, String? explicit) {
  if (explicit == null || explicit.isEmpty) return null;
  if (explicit.startsWith('$accountPart|')) return explicit;
  return null;
}

String deriveThreadKeyForStoredMessage({
  required String accountPart,
  required String fromDeviceId,
  String? toDeviceId,
  required String myDeviceId,
  String? explicitThreadKey,
}) {
  if (explicitThreadKey != null && explicitThreadKey.isNotEmpty) {
    return explicitThreadKey;
  }
  final to = toDeviceId;
  if (to != null &&
      to.isNotEmpty &&
      to != kS3VirtualDeviceId &&
      fromDeviceId != kS3VirtualDeviceId) {
    return threadKeyOneToOne(accountPart, fromDeviceId, to);
  }
  if (fromDeviceId != myDeviceId) {
    return threadKeyOneToOne(accountPart, fromDeviceId, myDeviceId);
  }
  return threadKeyLegacyBroadcast(accountPart);
}

File: web/src/lib/threadKey.ts
/** Same rules as `app/lib/chat/thread_key.dart` and backend `ThreadKeyUtil`. */

export const S3_VIRTUAL_DEVICE_ID = '__s3_cloud__';

const threadKindS3Cloud = 's3_cloud';
const threadKindLegacyBroadcast = 'legacy_broadcast';

export function accountPartLoggedIn(userId: string): string {
  return `u:${userId}`;
}

export function accountPartOffline(offlineUserId: string): string {
  return `o:${offlineUserId}`;
}

export function threadKeyOneToOne(accountPart: string, deviceIdA: string, deviceIdB: string): string {
  const a = deviceIdA <= deviceIdB ? deviceIdA : deviceIdB;
  const b = deviceIdA <= deviceIdB ? deviceIdB : deviceIdA;
  return `device|d1:${a}|d2:${b}`;
}

export function threadKeyS3Cloud(accountPart: string): string {
  return `${accountPart}|kind:${threadKindS3Cloud}`;
}

export function threadKeyLegacyBroadcast(accountPart: string): string {
  return `${accountPart}|kind:${threadKindLegacyBroadcast}`;
}

export function threadKeyForPeerSelection(params: {
  accountPart: string;
  myDeviceId: string;
  selectedPeerId: string;
}): string {
  if (params.selectedPeerId === S3_VIRTUAL_DEVICE_ID) {
    return threadKeyS3Cloud(params.accountPart);
  }
  return threadKeyOneToOne(params.accountPart, params.myDeviceId, params.selectedPeerId);
}

/** Logged-in web / Flutter online: build outbound envelope fields from sidebar selection. */
export function outboundForWebChat(
  userId: string,
  selectedPeerId: string | null,
  myDeviceId: string,
): { threadKey: string; toDeviceId?: string } | null {
  if (!selectedPeerId) return null;
  const accountPart = accountPartLoggedIn(userId);
  if (selectedPeerId === S3_VIRTUAL_DEVICE_ID) {
    return { threadKey: threadKeyS3Cloud(accountPart) };
  }
  return {
    threadKey: threadKeyOneToOne(accountPart, myDeviceId, selectedPeerId),
    toDeviceId: selectedPeerId,
  };
}

/** Unsigned-in 1:1 chat: device-send requires toDeviceId. */
export function outboundForGuestChat(
  selectedPeerId: string | null,
  myDeviceId: string,
): { threadKey: string; toDeviceId: string } | null {
  if (!selectedPeerId || selectedPeerId === S3_VIRTUAL_DEVICE_ID) return null;
  return {
    threadKey: threadKeyOneToOne(accountPartOffline(myDeviceId), myDeviceId, selectedPeerId),
    toDeviceId: selectedPeerId,
  };
}

export function threadKeyForS3WebPersist(
  userId: string,
  myDeviceId: string,
  toDeviceId: string | null | undefined,
): string {
  const ap = accountPartLoggedIn(userId);
  if (toDeviceId != null && toDeviceId !== '') {
    return threadKeyOneToOne(ap, myDeviceId, toDeviceId);
  }
  return threadKeyS3Cloud(ap);
}

export function deriveThreadKeyForStoredMessage(params: {
  accountPart: string;
  fromDeviceId: string;
  toDeviceId?: string | null;
  myDeviceId: string;
  explicitThreadKey?: string | null;
}): string {
  const { accountPart, fromDeviceId, myDeviceId, explicitThreadKey } = params;
  const to = params.toDeviceId;
  if (explicitThreadKey != null && explicitThreadKey.length > 0) {
    return explicitThreadKey;
  }
  if (
    to != null &&
    to !== '' &&
    to !== S3_VIRTUAL_DEVICE_ID &&
    fromDeviceId !== S3_VIRTUAL_DEVICE_ID
  ) {
    return threadKeyOneToOne(accountPart, fromDeviceId, to);
  }
  if (fromDeviceId !== myDeviceId) {
    return threadKeyOneToOne(accountPart, fromDeviceId, myDeviceId);
  }
  return threadKeyLegacyBroadcast(accountPart);
}
